src/toolkits/object_detection/one_shot_object_detection/one_shot_object_detector.cpp

```diff
@@ -133,85 +135,95 @@ gl_sframe augment_data(const gl_sframe &data,
   size_t total_augmented_rows = data.size() * backgrounds_size;
   table_printer table( { {"Images Augmented", 0}, {"Elapsed Time", 0}, {"Percent Complete", 0} } );
   if (verbose) {
-    logprogress_stream << "Augmenting input images using " << backgrounds.size() << " background images." << std::endl;
+    logprogress_stream << "Augmenting input images using " << backgrounds_size << " background images." << std::endl;
     table.print_header();
   }
-
+  std::vector<std::string> output_column_names = {"image", "annotation"};
+  std::vector<flex_type_enum> output_column_types = {flex_type_enum::IMAGE, flex_type_enum::DICT};
+  gl_sframe_writer output_writer(output_column_names, output_column_types);  
   auto column_index_map = generate_column_index_map(data.column_names());
-  std::vector<flexible_type> annotations;
-  std::vector<flexible_type> images;
-  int input_row_index = -1;
-  for (const auto& row: data.range_iterator()) {
-    input_row_index++;
-    flex_image object = row[column_index_map[image_column_name]].to<flex_image>();
-    std::string label = row[column_index_map[target_column_name]].to<flex_string>();
-    if (!(object.is_decoded())) {
-      decode_image_inplace(object);
-    }
-    int row_number = -1;
-    for (const auto& background_ft: backgrounds.range_iterator()) {
-      row_number++;
-      flex_image flex_background = background_ft.to<flex_image>();
-      if (!(flex_background.is_decoded())) {
-        decode_image_inplace(flex_background);
+  size_t nsegments = output_writer.num_segments();
+  size_t segment_size = backgrounds.size()/nsegments;
+  std::mutex counter_lock;
+  size_t augmented_counter = 0; // read and write protected by counter_lock mutex
+  
+  parallel_for(0, nsegments, [&](size_t segment_id){
+    auto segment_start = segment_id * segment_size;
+    auto segment_end = (segment_id+1) * segment_size;
+    for (size_t background_image_index = segment_start;
+                background_image_index < segment_end;
+                background_image_index++){
+      for (const auto& row: data.range_iterator()) {
+        flex_image object = row[column_index_map[image_column_name]].to<flex_image>();
+        std::string label = row[column_index_map[target_column_name]].to<flex_string>();
+        if (!(object.is_decoded())) {
+          decode_image_inplace(object);
+        }
+        auto background_ft = backgrounds[background_image_index];
```

Comment:
Retrieving each background by index one at a time is very inefficient. Use `gl_sarray::range_iterator` for this purpose. Note that you can specify the range of indices

Reply:
Done.